### PR TITLE
Modify the RSA_private_decrypt functions to check the padding in

### PR DESCRIPTION
**Pull request by bernd-edlinger**
constant time with a memory access pattern that does not depend
on secret information.

### CODE REVIEW

File: crypto/rsa/rsa_oaep.c [hunk 1]
```diff
@@ -234,25 +234,21 @@ int RSA_padding_check_PKCS1_OAEP_mgf1(unsigned char *to, int tlen,
...
                                     dblen - mdlen - 1, tlen);
-    msg_index = constant_time_select_int(good, msg_index, dblen - tlen);
-    mlen = dblen - msg_index;
-    for (mask = good, i = 0; i < tlen; i++) {
-        unsigned int equals = constant_time_eq(msg_index, dblen);
-
-        msg_index -= tlen & equals;  /* rewind at EOF */
-        mask &= ~equals;  /* mask = 0 at EOF */
-        to[i] = constant_time_select_8(mask, db[msg_index++], to[i]);
+    for (msg_index = 1; msg_index < dblen - mdlen - 1; msg_index <<= 1) {
```
Comment (paulidale) [inline crypto/rsa/rsa_oaep.c]: Is there a more detailed description about how this pattern works?
It is not obvious it does what it does.  It appears to be copying over the destination multiple times which is odd.

If not, perhaps a comment describing it?

Comment (paulidale) [reply crypto/rsa/rsa_oaep.c]: I figured it out the second time around.
It still wasn't obvious.

Comment (bernd-edlinger) [reply crypto/rsa/rsa_oaep.c]: Okay, will try to come up with a comment explaining the idea.

Comment (paulidale) [reply crypto/rsa/rsa_oaep.c]: A clever solution to a difficult problem :)

Comment (bernd-edlinger) [reply crypto/rsa/rsa_oaep.c]: I admit I have stolen the idea from @davidben

Comment (paulidale) [reply crypto/rsa/rsa_oaep.c]: How about something along this line:

"Copy the memory back in a way that does reveal the size of the data being copied via a timing side channel.  The requires copying parts of the buffer multiple times based on the bits set in the real length.  Clear bits do a non-copy with similar access pattern."

**Comment by bernd-edlinger** [comment]
How about adding that to the comment above:

The loop below combines conditinal moves by 2^X bytes. Each move
is enabled or disabled dependent on the bit set in the required
displacement. It has a complexity of O(N*log(N))).

File: crypto/rsa/rsa_oaep.c (hunk 1 above)
Comment (bernd-edlinger) [reply crypto/rsa/rsa_oaep.c]: opps, I did not see your comment above.  I guess I take yours and add the statement about complexity.

Comment (bernd-edlinger) [reply crypto/rsa/rsa_oaep.c]: I guess I can say "Clear bits do a non-copy with identical access pattern".

File: crypto/rsa/rsa_oaep.c [hunk 2]
```diff
@@ -234,25 +234,24 @@ int RSA_padding_check_PKCS1_OAEP_mgf1(unsigned char *to, int tlen,
...
-     * where |mlen'| is "saturated" |mlen| value. Deducing information
-     * about failure or |mlen| would take attacker's ability to observe
-     * memory access pattern with byte granularity *as it occurs*. It
-     * should be noted that failure is indistinguishable from normal
-     * operation if |tlen| is fixed by protocol.
+     * Move the result in-place by |dblen|-|mdlen|-1-|mlen| bytes to the left.
+     * Then if |good| move |mlen| bytes from |db|+|mdlen|+1 to |to|.
+     * Otherwise leave |to| unchanged.
+     * Do that all in constant time.
+     * The loop below combines conditional moves by 2^X bytes. Each move
```
Comment (paulidale) [inline crypto/rsa/rsa_oaep.c]: _of_ instead of _by_?

File: crypto/rsa/rsa_oaep.c (hunk 1 above)
Comment (paulidale) [reply crypto/rsa/rsa_oaep.c]: You statement is good enough I think.  It makes it clear that it is intentional and gives enough information to work out what it is doing.

File: crypto/rsa/rsa_oaep.c (hunk 2 above)
Comment (bernd-edlinger) [reply crypto/rsa/rsa_oaep.c]: I dropped that sentence.

**Comment by bernd-edlinger** [comment]
Sorry, messages crossed.
I think I should say "it has an overall complexity" instead of "a" right?

File: crypto/rsa/rsa_oaep.c [hunk 3]
```diff
@@ -234,25 +234,25 @@ int RSA_padding_check_PKCS1_OAEP_mgf1(unsigned char *to, int tlen,
...
-     * where |mlen'| is "saturated" |mlen| value. Deducing information
-     * about failure or |mlen| would take attacker's ability to observe
-     * memory access pattern with byte granularity *as it occurs*. It
-     * should be noted that failure is indistinguishable from normal
-     * operation if |tlen| is fixed by protocol.
+     * Move the result in-place by |dblen|-|mdlen|-1-|mlen| bytes to the left.
+     * Then if |good| move |mlen| bytes from |db|+|mdlen|+1 to |to|.
+     * Otherwise leave |to| unchanged.
+     * Copy the memory back in a way that does not reveal the size of
+     * the data being copied via a timing side channel. The requires copying
```
Comment (paulidale) [inline crypto/rsa/rsa_oaep.c]: **This** requires copying

**Comment by paulidale** [comment]
yes, drop the "a".

File: crypto/rsa/rsa_oaep.c (hunk 3 above)
Comment (paulidale) [reply crypto/rsa/rsa_oaep.c]: My mistake, not yours.

Comment (bernd-edlinger) [reply crypto/rsa/rsa_oaep.c]: NP. already fixed.

**Comment by bernd-edlinger** [comment]
> yes, drop the "a".

good point. Thanks.

**Comment by bernd-edlinger** [comment]
Merged to all active branches.  Thanks!

**Pull request closed without merging**